WordPress/Classes/ViewRelated/Support/SupportTableViewController.swift

```diff
@@ -278,11 +279,13 @@ private extension SupportTableViewController {
         let title: String
         let showIndicator: Bool
         let action: ImmuTableAction?
+        let accessibilityIdentifier: String?
 
-        init(title: String, action: @escaping ImmuTableAction, showIndicator: Bool = false) {
+        init(title: String, action: @escaping ImmuTableAction, showIndicator: Bool = false, accessibilityIdentifier: String? = nil) {
```

Comment:
Thanks! Credit goes to @leandroalonso who showed me how to do this:
p1625472413270600-slack-C011BKNU1V5

Looking back over this, I realize that `HelpRow` has an `init()` with those default parameters, while `SupportEmailRow` below (line 302) only has the `let`s and `configureCell()`.
Do you see a clear reason `SupportEmailRow` doesn't use an `init()`?

Reply:
> Do you see a clear reason `SupportEmailRow` doesn't use an `init()`?

I think your question was more about why `SupportEmailRow` doesn't have an `init` method in which it defines some default values, like `HelpRow` does. But, just in case, I'd like to point out that `SupportEmailRow` _has_ an `init` method (although, it's not `init()` which technically speaking would mean an init without arguments, but would be written as `init(title:,value:,accessibilityHint:,action:)`)

The reason it's not explicitly defined is that the Swift compiler synthesizes initializer for `struct`s if no initializer is defined. This is called [memberwise initializer](https://docs.swift.org/swift-book/LanguageGuide/Initialization.html)

> Structure types automatically receive a memberwise initializer if they don’t define any of their own custom initializers

<img width="1075" alt="Screen Shot 2021-07-19 at 4 11 23 pm" src="https://user-images.githubusercontent.com/1218433/126111612-8d784ccc-7fea-4e8f-b94a-f26ea69848a9.png">

That's a super useful feature that makes writing and using straightforward code seamless. 🤓 

I don't have an answer to why `SupportEmailRow` doesn't define some defaults, while `HelpRow` does. But maybe the the code history has something say about it ☝️ 

All our code is stored in Git, a version control system (VCS). Developer track discrete changes to the code in the form of commits, and add a title to each commit. There are known best practices on how to write good commit messages, that can help future readers learn about the rational behind a change. 

Git has an ill named command called `git blame` that shows the commit that last changed each line of code. Xcode expose this functionality through the "Authors" UI:

<img width="236" alt="Screen Shot 2021-07-19 at 4 24 44 pm" src="https://user-images.githubusercontent.com/1218433/126112537-6ea1afeb-08e2-4c4c-be18-cffb9e6fd7c5.png">

I find, though, that the tool that offers the best ergonomics to go spelunking through the code's history is GitHub. We can open `SupportTableViewController.swift` in GitHub and select "Blame":

<img width="1399" alt="Screen Shot 2021-07-19 at 4 26 53 pm" src="https://user-images.githubusercontent.com/1218433/126112932-786bf993-2f55-4f00-a00c-5c7030be6b90.png">

The blame view shows the commit that last changed each line, as well as a button to scroll back through the previous versions:

<img width="1733" alt="Screen Shot 2021-07-19 at 4 29 48 pm" src="https://user-images.githubusercontent.com/1218433/126113086-55d98883-d64f-492c-8382-819c5f425e74.png">

Using that, we can learn that `HelpRow` was added [3+ years ago](https://github.com/wordpress-mobile/WordPress-iOS/commit/6f23bc34ef52ebe1064c1a1a0edcf38c2a8b7c5f#diff-49dc6dec4a731852be5a9f38c831ff0ed74c648e5a412aafd19404fcbcb0bbcaR133-R149) and already had that explicit init. There's no explanation in the commit message as of why it was written like that, but we could ask the author, @ScoutHarris if she remembers 😄 

The only thing I notice is that the `action` property is defined as wrapped in an `Optional`, [as required by the `ImmuTableRow` protocol](https://github.com/wordpress-mobile/WordPress-iOS/blob/17488be698b505545d953c02c20476fb5555c92d/WordPress/Classes/Utility/ImmuTable.swift#L143), but it is not in the `init`. My guess without any extra context would be that @ScoutHarris didn't want to allow callers of `HelpRow` to app a `nil` (i.e. `Optional<ImmuTableAction>.none`) value.

I decided to look into `HelpRow` instead of `SupportEmailRow` because, since the memberwise initializer was introduced in an early version of Swift, it's usually always used instead of explicitly defining one (why doing more work than necessary, right?) so I was more curious about why `HelpRow` had a custom one.

There's a lot more to say about Swift initialization and searching information in Git, but I think I've pontificated long enough for as single comment 😅 

---

Further reading on how to write good commit messages and why it matters:

- https://thoughtbot.com/blog/5-useful-tips-for-a-better-commit-message
- https://mislav.net/2014/02/hidden-documentation
- https://mokacoding.com/blog/your-git-log-should-tell-a-story/
- https://mokacoding.com/blog/better-merging-for-github-pull-requests/

Reply:
> I think your question was more about why SupportEmailRow doesn't have an init method in which it defines some default values, like HelpRow does.

Yep, completely right. I wasn't aware of memberwise initializers, so thanks for pointing that out!

TIL about that "Authors" view as well, though I've used the "history" in the inspectors sidebar to help me find that info in the past. I totally agree that GitHub's _blame_ is way easier to use.

> I decided to look into HelpRow instead of SupportEmailRow because, since the memberwise initializer was introduced in an early version of Swift, it's usually always used instead of explicitly defining one (why doing more work than necessary, right?) so I was more curious about why HelpRow had a custom one.

Ok, that make sense. If I understand, the memberwise initializer will use all the properties defined in the struct (so in this case, `title`, `value`, `accessibilityHint`, `action`, and now `accessibilityIdentifier`). So, unless you have a reason to define custom initializer, you can generally rely on memberwise initialization without any trouble.
Am I getting that right?

Thanks!

Reply:
> HelpRow was added 3+ years ago and already had that explicit init. 

I _think_ I added an explicit `init` so `showIndicator` could have a default value. Only one row has the indicator, so having it passed for the others was superfluous. 

> The only thing I notice is that the action property is defined as wrapped in an Optional, as required by the ImmuTableRow protocol, but it is not in the init.

This looks like an oversight. Although it does work without `action` being explicitly included in the `init` params, which is probably why I didn't notice.

Reply:
@thehenrybyrd 

> So, unless you have a reason to define custom initializer, you can generally rely on memberwise initialization without any trouble.
> Am I getting that right?

Yes.